contracts/src/utils/cryptography/ecdsa.rs

```diff
@@ -0,0 +1,248 @@
+//! Elliptic Curve Digital Signature Algorithm (ECDSA) operations.
+//!
+//! These functions can be used to verify that a message was signed
+//! by the holder of the private keys of a given address.
+use alloc::vec::Vec;
+
+use alloy_primitives::{address, uint, Address, B256, U256};
+use alloy_sol_types::{sol, SolType};
+use stylus_proc::SolidityError;
+use stylus_sdk::{
+    call::{self, Call},
+    storage::TopLevelStorage,
+};
+
+const ECRECOVER_ADDR: Address =
+    address!("0000000000000000000000000000000000000001");
+
+const EIP2_VALUE: U256 = uint!(
+    0x7FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF5D576E7357A4501DDFE92F46681B20A0_U256
+);
+
+sol! {
+    /// The signature derives the `Address::ZERO`.
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+    error ECDSAInvalidSignature();
+
+    /// The signature has an invalid length.
+    ///
+    /// * `length` - Length of the signature.
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+    error ECDSAInvalidSignatureLength(uint256 length);
+
+    /// The signature has an `S` value that is in the upper half order.
+    ///
+    /// * `s` - Invalid `S` value.
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+    error ECDSAInvalidSignatureS(bytes32 s);
+}
+
+/// An error that occurred in the implementation of an `ECDSA` library.
+#[derive(SolidityError, Debug)]
+pub enum Error {
+    /// The signature derives the `Address::ZERO`.
+    InvalidSignature(ECDSAInvalidSignature),
+    /// The signature has an invalid length.
+    InvalidSignatureLength(ECDSAInvalidSignatureLength),
+    /// The signature has an `S` value that is in the upper half order.
+    InvalidSignatureS(ECDSAInvalidSignatureS),
+}
+
+sol! {
+    /// Struct with callable data to the `ecrecover` precompile.
+    #[allow(missing_docs)]
+    struct ECRecoverData {
+        /// EIP-191 Hash of the message.
+        bytes32 hash;
+        /// `v` value from the signature.
+        uint8 v;
+        /// `r` value from the signature.
+        bytes32 r;
+        /// `s` value from the signature.
+        bytes32 s;
+    }
+}
+
+/// Returns the address that signed a hashed message (`hash`).
+///
+/// # Arguments
+///
+/// * `storage` - Write access to storage.
+/// * `hash` - Hash of the message.
+/// * `v` - `v` value from the signature.
+/// * `r` - `r` value from the signature.
+/// * `s` - `s` value from the signature.
+///
+/// # Errors
+///
+/// * If the `s` value is grater than `EIP2_VALUE`, then the error
+/// [`Error::InvalidSignatureS`] is returned.
+/// * If the recovered address is `Address::ZERO`, then the error
+/// [`Error::InvalidSignature`] is returned.
+///
+/// # Panics
+///
+/// * If `ecrecover` precompile fail to execute.
+pub fn recover(
+    storage: &mut impl TopLevelStorage,
+    hash: B256,
+    v: u8,
+    r: B256,
+    s: B256,
+) -> Result<Address, Error> {
+    validate_s_value(&s)?;
+    // If the signature is valid (and not malleable), return the signer address.
+    evm_recover(storage, hash, v, r, s)
```

Comment:
I'd rather rename this to something else, like `_recover`.

I would also inline it, but I don't feel strongly about it, so feel free to ignore.